test_config unit tests fail for different cluster config files

## Background

Three unit tests in `test_config.py` fail when the cluster config file `prod.json` doesn't exist.

In my setup, the production cluster is named `production.json`. The contents of the cluster config are supposed to be mocked in `test_config.py`, but `CLIConfig()` will list the actual contents of the `conf/clusters` directory to retrieve the cluster filenames.

Subsequently `CLIConfig` attempts to open `conf/clusters/production.json` and doesn't get the mocked content (or mocked `prod` cluster name) leading to KeyErrors.

```
======================================================================
ERROR: CLI - Adding CloudWatch metric alarm, cluster
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jeremy/.local/share/virtualenvs/streamalert-iKoON-og/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jeremy/.local/share/virtualenvs/streamalert-iKoON-og/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/Users/jeremy/projects/streamalert/tests/unit/stream_alert_cli/test_config.py", line 98, in test_cluster_alarm_creation
    config.add_metric_alarm(alarm_info)
  File "/Users/jeremy/projects/streamalert/stream_alert_cli/config.py", line 366, in add_metric_alarm
    self._add_metric_alarm_per_cluster(alarm_info, metric_function)
  File "/Users/jeremy/projects/streamalert/stream_alert_cli/config.py", line 205, in _add_metric_alarm_per_cluster
    function_config = (self.config['clusters'][cluster]['modules']
KeyError: 'prod'

======================================================================
ERROR: CLI - Metric toggling
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jeremy/.local/share/virtualenvs/streamalert-iKoON-og/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jeremy/.local/share/virtualenvs/streamalert-iKoON-og/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/Users/jeremy/projects/streamalert/tests/unit/stream_alert_cli/test_config.py", line 62, in test_toggle_metric
    config.toggle_metrics(True, ['prod'], ['alert_processor'])
  File "/Users/jeremy/projects/streamalert/stream_alert_cli/config.py", line 161, in toggle_metrics
    [function]['enable_metrics'] = enabled
KeyError: 'prod'

======================================================================
FAIL: CLI - Aggregate alarm check
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jeremy/.local/share/virtualenvs/streamalert-iKoON-og/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/jeremy/projects/streamalert/tests/unit/stream_alert_cli/test_config.py", line 77, in test_cluster_alarm_exists
    assert_true(result)
AssertionError: False is not true

----------------------------------------------------------------------
Ran 353 tests in 68.910s

FAILED (errors=2, failures=1)
```

## Desired Change

Create a mock filesystem using pyfakefs instead of only mocking out `__builtin__.open()`, so that `os.path.listdir()` will return `conf/clusters/prod.json`.

I'm happy to submit a PR for this.

Hi @stoggi - great catch, thanks for this!! You're welcome to submit a PR if you have a fix in mind (which it sounds like you do). If not, we can try to get a fix in place asap. Just let us know if you decide to work on it!

> On Oct 18, 2017, at 9:24 PM, Jeremy Stott <notifications@github.com> wrote:
> 
> Background
> 
> Three unit tests in test_config.py fail when the cluster config file prod.json doesn't exist.
> 
> In my setup, the production cluster is named production.json. The contents of the cluster config are supposed to be mocked in test_config.py, but CLIConfig() will list the actual contents of the conf/clusters directory to retrieve the cluster filenames.
> 
> Subsequently CLIConfig attempts to open conf/clusters/production.json and doesn't get the mocked content (or mocked prod cluster name) leading to KeyErrors.
> 
> ======================================================================
> ERROR: CLI - Adding CloudWatch metric alarm, cluster
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/jeremy/.local/share/virtualenvs/streamalert-iKoON-og/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
>     self.test(*self.arg)
>   File "/Users/jeremy/.local/share/virtualenvs/streamalert-iKoON-og/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
>     return func(*args, **keywargs)
>   File "/Users/jeremy/projects/streamalert/tests/unit/stream_alert_cli/test_config.py", line 98, in test_cluster_alarm_creation
>     config.add_metric_alarm(alarm_info)
>   File "/Users/jeremy/projects/streamalert/stream_alert_cli/config.py", line 366, in add_metric_alarm
>     self._add_metric_alarm_per_cluster(alarm_info, metric_function)
>   File "/Users/jeremy/projects/streamalert/stream_alert_cli/config.py", line 205, in _add_metric_alarm_per_cluster
>     function_config = (self.config['clusters'][cluster]['modules']
> KeyError: 'prod'
> 
> ======================================================================
> ERROR: CLI - Metric toggling
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/jeremy/.local/share/virtualenvs/streamalert-iKoON-og/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
>     self.test(*self.arg)
>   File "/Users/jeremy/.local/share/virtualenvs/streamalert-iKoON-og/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
>     return func(*args, **keywargs)
>   File "/Users/jeremy/projects/streamalert/tests/unit/stream_alert_cli/test_config.py", line 62, in test_toggle_metric
>     config.toggle_metrics(True, ['prod'], ['alert_processor'])
>   File "/Users/jeremy/projects/streamalert/stream_alert_cli/config.py", line 161, in toggle_metrics
>     [function]['enable_metrics'] = enabled
> KeyError: 'prod'
> 
> ======================================================================
> FAIL: CLI - Aggregate alarm check
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/Users/jeremy/.local/share/virtualenvs/streamalert-iKoON-og/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
>     self.test(*self.arg)
>   File "/Users/jeremy/projects/streamalert/tests/unit/stream_alert_cli/test_config.py", line 77, in test_cluster_alarm_exists
>     assert_true(result)
> AssertionError: False is not true
> 
> ----------------------------------------------------------------------
> Ran 353 tests in 68.910s
> 
> FAILED (errors=2, failures=1)
> Desired Change
> 
> Create a mock filesystem using pyfakefs instead of only mocking out __builtin__.open(), so that os.path.listdir() will return conf/clusters/prod.json.
> 
> I'm happy to submit a PR for this.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

--- a/requirements.txt
+++ b/requirements.txt
@@ -45,6 +45,7 @@ ply==3.10
 psutil==5.1.3
 pyasn1==0.2.3
 pycodestyle==2.3.1
+pyfakefs==3.2
 pyflakes==1.5.0
 Pygments==2.2.0
 pylint==1.7.2

--- a/tests/unit/stream_alert_cli/test_config.py
+++ b/tests/unit/stream_alert_cli/test_config.py
@@ -14,67 +14,73 @@ See the License for the specific language governing permissions and
 limitations under the License.
 """
 # pylint: disable=protected-access
-from contextlib import nested
 import json
 
 from mock import Mock, patch
 from nose.tools import assert_equal, assert_true
+from pyfakefs import fake_filesystem_unittest
 
 from stream_alert_cli.config import CLIConfig
-from tests.unit.helpers.base import basic_streamalert_config, mock_open
+from tests.unit.helpers.base import basic_streamalert_config
 
 
 class TestCLIConfig(object):
     """Test class for CLIConfig"""
+    # pylint: disable=no-self-use
 
     def __init__(self):
-        self.mocked_opens = None
+        self.fs_patcher = None
 
     def setup(self):
         """Setup before each method"""
         config_data = basic_streamalert_config()
-        self.mocked_opens = [
-            mock_open('conf/global.json', json.dumps(config_data['global'])),
-            mock_open('conf/lambda.json', json.dumps(config_data['lambda'])),
-            mock_open('conf/clusters/prod.json', json.dumps(config_data['clusters']['prod']))
-        ]
+
+        self.fs_patcher = fake_filesystem_unittest.Patcher()
+        self.fs_patcher.setUp()
+
+        self.fs_patcher.fs.CreateFile('/conf/global.json',
+                                      contents=json.dumps(config_data['global']))
+        self.fs_patcher.fs.CreateFile('/conf/lambda.json',
+                                      contents=json.dumps(config_data['lambda']))
+        self.fs_patcher.fs.CreateFile('/conf/clusters/prod.json',
+                                      contents=json.dumps(config_data['clusters']['prod']))
+
+    def teardown(self):
+        """Teardown after each method"""
+        self.fs_patcher.tearDown()
 
     def test_load_config(self):
         """CLI - Load config"""
-        with nested(*self.mocked_opens):
-            config = CLIConfig()
-            assert_equal(config['global']['account']['prefix'], 'unit-testing')
+        config = CLIConfig()
+        assert_equal(config['global']['account']['prefix'], 'unit-testing')
 
     @patch('logging.Logger.error')
     @patch('stream_alert_cli.config.CLIConfig.write')
     def test_toggle_metric(self, write_mock, log_mock):
         """CLI - Metric toggling"""
-        with nested(*self.mocked_opens):
-            config = CLIConfig()
+        config = CLIConfig()
 
-            config.toggle_metrics(True, [], ['athena_partition_refresh'])
-            write_mock.assert_called()
+        config.toggle_metrics(True, [], ['athena_partition_refresh'])
+        write_mock.assert_called()
 
-            del config.config['lambda']['athena_partition_refresh_config']
-            config.toggle_metrics(True, [], ['athena_partition_refresh'])
-            log_mock.assert_called_with('No Athena configuration found; please initialize first.')
+        del config.config['lambda']['athena_partition_refresh_config']
+        config.toggle_metrics(True, [], ['athena_partition_refresh'])
+        log_mock.assert_called_with('No Athena configuration found; please initialize first.')
 
-            config.toggle_metrics(True, ['prod'], ['alert_processor'])
-            write_mock.assert_called()
+        config.toggle_metrics(True, ['prod'], ['alert_processor'])
+        write_mock.assert_called()
 
     def test_aggregate_alarm_exists(self):
         """CLI - Aggregate alarm check"""
-        with nested(*self.mocked_opens):
-            config = CLIConfig()
-            result = config._alarm_exists('Aggregate Unit Testing Failed Parses Alarm')
-            assert_true(result)
+        config = CLIConfig()
+        result = config._alarm_exists('Aggregate Unit Testing Failed Parses Alarm')
+        assert_true(result)
 
     def test_cluster_alarm_exists(self):
         """CLI - Aggregate alarm check"""
-        with nested(*self.mocked_opens):
-            config = CLIConfig()
-            result = config._alarm_exists('Prod Unit Testing Failed Parses Alarm')
-            assert_true(result)
+        config = CLIConfig()
+        result = config._alarm_exists('Prod Unit Testing Failed Parses Alarm')
+        assert_true(result)
 
     @patch('stream_alert_cli.config.CLIConfig.write', Mock())
     @patch('logging.Logger.info')
@@ -93,14 +99,13 @@ class TestCLIConfig(object):
             'comparison_operator': 'LessThanThreshold'
         }
 
-        with nested(*self.mocked_opens):
-            config = CLIConfig()
-            config.add_metric_alarm(alarm_info)
-            log_mock.assert_called_with('Successfully added \'%s\' metric alarm for the '
-                                        '\'%s\' function to \'conf/clusters/%s.json\'.',
-                                        'Prod Unit Testing Total Records Alarm',
-                                        'rule_processor',
-                                        'prod')
+        config = CLIConfig()
+        config.add_metric_alarm(alarm_info)
+        log_mock.assert_called_with('Successfully added \'%s\' metric alarm for the '
+                                    '\'%s\' function to \'conf/clusters/%s.json\'.',
+                                    'Prod Unit Testing Total Records Alarm',
+                                    'rule_processor',
+                                    'prod')
 
     @patch('stream_alert_cli.config.CLIConfig.write', Mock())
     @patch('logging.Logger.info')
@@ -119,9 +124,8 @@ class TestCLIConfig(object):
             'comparison_operator': 'LessThanThreshold'
         }
 
-        with nested(*self.mocked_opens):
-            config = CLIConfig()
-            config.add_metric_alarm(alarm_info)
-            log_mock.assert_called_with('Successfully added \'%s\' metric alarm to '
-                                        '\'conf/global.json\'.',
-                                        'Aggregate Unit Testing Total Records Alarm')
+        config = CLIConfig()
+        config.add_metric_alarm(alarm_info)
+        log_mock.assert_called_with('Successfully added \'%s\' metric alarm to '
+                                    '\'conf/global.json\'.',
+                                    'Aggregate Unit Testing Total Records Alarm')

--- a/tests/unit/stream_alert_cli/test_config.py
+++ b/tests/unit/stream_alert_cli/test_config.py
@@ -14,7 +14,6 @@ See the License for the specific language governing permissions and
 limitations under the License.
 """
 # pylint: disable=protected-access
-# pylint: disable=no-self-use
 import json
 
 from mock import Mock, patch

--- a/tests/unit/stream_alert_cli/test_config.py
+++ b/tests/unit/stream_alert_cli/test_config.py
@@ -14,6 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 """
 # pylint: disable=protected-access
+# pylint: disable=no-self-use
 import json
 
 from mock import Mock, patch
@@ -26,9 +27,9 @@ from tests.unit.helpers.base import basic_streamalert_config
 
 class TestCLIConfig(object):
     """Test class for CLIConfig"""
-    # pylint: disable=no-self-use
 
     def __init__(self):
+        self.config = None
         self.fs_patcher = None
 
     def setup(self):
@@ -45,41 +46,40 @@ class TestCLIConfig(object):
         self.fs_patcher.fs.CreateFile('/conf/clusters/prod.json',
                                       contents=json.dumps(config_data['clusters']['prod']))
 
+        # Create the config instance after creating the fake filesystem so that
+        # CLIConfig uses our mocked config files instead of the real ones.
+        self.config = CLIConfig()
+
     def teardown(self):
         """Teardown after each method"""
         self.fs_patcher.tearDown()
 
     def test_load_config(self):
         """CLI - Load config"""
-        config = CLIConfig()
-        assert_equal(config['global']['account']['prefix'], 'unit-testing')
+        assert_equal(self.config['global']['account']['prefix'], 'unit-testing')
 
     @patch('logging.Logger.error')
     @patch('stream_alert_cli.config.CLIConfig.write')
     def test_toggle_metric(self, write_mock, log_mock):
         """CLI - Metric toggling"""
-        config = CLIConfig()
-
-        config.toggle_metrics(True, [], ['athena_partition_refresh'])
+        self.config.toggle_metrics(True, [], ['athena_partition_refresh'])
         write_mock.assert_called()
 
-        del config.config['lambda']['athena_partition_refresh_config']
-        config.toggle_metrics(True, [], ['athena_partition_refresh'])
+        del self.config.config['lambda']['athena_partition_refresh_config']
+        self.config.toggle_metrics(True, [], ['athena_partition_refresh'])
         log_mock.assert_called_with('No Athena configuration found; please initialize first.')
 
-        config.toggle_metrics(True, ['prod'], ['alert_processor'])
+        self.config.toggle_metrics(True, ['prod'], ['alert_processor'])
         write_mock.assert_called()
 
     def test_aggregate_alarm_exists(self):
         """CLI - Aggregate alarm check"""
-        config = CLIConfig()
-        result = config._alarm_exists('Aggregate Unit Testing Failed Parses Alarm')
+        result = self.config._alarm_exists('Aggregate Unit Testing Failed Parses Alarm')
         assert_true(result)
 
     def test_cluster_alarm_exists(self):
         """CLI - Aggregate alarm check"""
-        config = CLIConfig()
-        result = config._alarm_exists('Prod Unit Testing Failed Parses Alarm')
+        result = self.config._alarm_exists('Prod Unit Testing Failed Parses Alarm')
         assert_true(result)
 
     @patch('stream_alert_cli.config.CLIConfig.write', Mock())
@@ -99,8 +99,7 @@ class TestCLIConfig(object):
             'comparison_operator': 'LessThanThreshold'
         }
 
-        config = CLIConfig()
-        config.add_metric_alarm(alarm_info)
+        self.config.add_metric_alarm(alarm_info)
         log_mock.assert_called_with('Successfully added \'%s\' metric alarm for the '
                                     '\'%s\' function to \'conf/clusters/%s.json\'.',
                                     'Prod Unit Testing Total Records Alarm',
@@ -124,8 +123,7 @@ class TestCLIConfig(object):
             'comparison_operator': 'LessThanThreshold'
         }
 
-        config = CLIConfig()
-        config.add_metric_alarm(alarm_info)
+        self.config.add_metric_alarm(alarm_info)
         log_mock.assert_called_with('Successfully added \'%s\' metric alarm to '
                                     '\'conf/global.json\'.',
                                     'Aggregate Unit Testing Total Records Alarm')

